packages/alfa-network/src/network.ts

```diff
@@ -0,0 +1,317 @@
+import { Equatable } from "@siteimprove/alfa-equatable";
+import { Graph } from "@siteimprove/alfa-graph";
+import { Hashable, Hash } from "@siteimprove/alfa-hash";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import { Serializable } from "@siteimprove/alfa-json";
+import { Map } from "@siteimprove/alfa-map";
+import { Sequence } from "@siteimprove/alfa-sequence";
+import { Set } from "@siteimprove/alfa-set";
+
+export class Network<N, E>
+  implements
+    Iterable<[N, Iterable<[N, Iterable<E>]>]>,
+    Equatable,
+    Hashable,
+    Serializable<Network.JSON<N, E>> {
+  public static of<N, E>(nodes: Map<N, Map<N, Set<E>>>): Network<N, E> {
+    return new Network(nodes);
+  }
+
+  private static _empty = new Network<never, never>(Map.empty());
+
+  public static empty<N, E>(): Network<N, E> {
+    return this._empty;
+  }
+
+  private readonly _nodes: Map<N, Map<N, Set<E>>>;
+
+  private constructor(nodes: Map<N, Map<N, Set<E>>>) {
+    this._nodes = nodes;
+  }
+
+  public get size(): number {
+    return this._nodes.size;
+  }
+
+  public nodes(): Iterable<N> {
+    return this._nodes.keys();
+  }
+
+  public neighbors(node: N): Iterable<[N, Iterable<E>]> {
+    return this._nodes.get(node).getOr([]);
+  }
+
+  public has(node: N): boolean {
+    return this._nodes.has(node);
+  }
+
+  public add(node: N): Network<N, E> {
+    if (this.has(node)) {
+      return this;
+    }
+
+    return new Network(this._nodes.set(node, Map.empty()));
+  }
+
+  public delete(node: N): Network<N, E> {
+    let nodes = this._nodes;
+
+    if (!nodes.has(node)) {
+      return this;
+    }
+
+    return new Network(
+      nodes.delete(node).map((neighbors) => neighbors.delete(node))
+    );
+  }
+
+  public connect(from: N, to: N, edge: E): Network<N, E> {
+    let nodes = this._nodes;
+
+    if (!nodes.has(from)) {
+      nodes = nodes.set(from, Map.empty());
+    }
+
+    if (!nodes.has(to)) {
+      nodes = nodes.set(to, Map.empty());
+    }
+
+    return new Network(
+      nodes.set(
+        from,
+        nodes
+          .get(from)
+          .map((from) =>
+            from.set(
+              to,
+              from
+                .get(to)
+                .map((edges) => edges.add(edge))
+                .getOrElse(() => Set.of(edge))
+            )
+          )
+          .get()
+      )
+    );
+  }
+
+  public disconnect(from: N, to: N): Network<N, E>;
+
+  public disconnect(from: N, to: N, edge: E): Network<N, E>;
```

Comment:
This could also take a set of edge to disconnect several but not all in one go.

Reply:
Good idea! 👍